File: StopMotionApp/Sources/StopMotionKit/FCPXMLBuilder.swift
import Foundation

/// A still image placed on the Final Cut Pro timeline.
public struct FCPXMLStill: Equatable, Sendable {
    public var url: URL
    public var width: Int
    public var height: Int

    public init(url: URL, width: Int, height: Int) {
        self.url = url
        self.width = width
        self.height = height
    }
}

/// Builds an FCPXML 1.10 document (Final Cut Pro 10.6 and later).
///
/// Each still sits on the primary storyline for `timing.framesPerImage` frames and, when
/// `timing.crossfadeFrames` is non-zero, a Cross Dissolve is centred on every cut. The output
/// matches `ImgConcat`'s `FcpxmlBuilder`, so both tools produce the same timeline.
public struct FCPXMLBuilder: Sendable {
    public static let version = "1.10"
    /// Final Cut Pro's built-in Cross Dissolve transition.
    public static let crossDissolveUID = "FxPlug:4731E73A-8DAC-4113-9A30-AE85B1761265"

    public var projectName: String
    public var eventName: String
    public var width: Int
    public var height: Int
    public var timing: FrameTiming

    public init(projectName: String, eventName: String = "StopMotion", width: Int, height: Int, timing: FrameTiming) {
        self.projectName = projectName
        self.eventName = eventName
        self.width = width
        self.height = height
        self.timing = timing
    }

    public enum BuildError: Error, Equatable {
        case noStills
    }

    public func build(stills: [FCPXMLStill]) throws -> String {
        guard !stills.isEmpty else { throw BuildError.noStills }

        let fps = timing.frameRate
        let perImage = timing.framesPerImage
        let crossfade = timing.crossfadeFrames

        var resources: [String] = []
        let sequenceFormatID = "r1"
        resources.append(
            #"<format id="\#(sequenceFormatID)" name="\#(Self.escape(Self.sequenceFormatName(width: width, height: height, fps: fps)))" frameDuration="\#(Self.time(frames: 1, fps: fps))" width="\#(width)" height="\#(height)" colorSpace="1-1-1 (Rec. 709)"/>"#
        )

        var nextID = 2
        var transitionEffectID: String?
        if crossfade > 0 && stills.count > 1 {
            let id = "r\(nextID)"
            nextID += 1
            transitionEffectID = id
            resources.append(#"<effect id="\#(id)" name="Cross Dissolve" uid="\#(Self.crossDissolveUID)"/>"#)
        }

        // Stills share a rate-undefined format per distinct size.
        var stillFormats: [String: String] = [:]
        var assetIDs: [String] = []
        for still in stills {
            let sizeKey = "\(still.width)x\(still.height)"
            let formatID: String
            if let existing = stillFormats[sizeKey] {
                formatID = existing
            } else {
                formatID = "r\(nextID)"
                nextID += 1
                stillFormats[sizeKey] = formatID
                resources.append(
                    #"<format id="\#(formatID)" name="FFVideoFormatRateUndefined" width="\#(still.width)" height="\#(still.height)" colorSpace="1-13-1"/>"#
                )
            }

            let assetID = "r\(nextID)"
            nextID += 1
            assetIDs.append(assetID)
            let name = Self.escape(still.url.deletingPathExtension().lastPathComponent)
            let src = Self.escape(still.url.standardizedFileURL.absoluteString)
            resources.append(
                """
                <asset id="\(assetID)" name="\(name)" start="0s" duration="0s" hasVideo="1" format="\(formatID)" videoSources="1">
                    <media-rep kind="original-media" src="\(src)"/>
                </asset>
                """
            )
        }

        var spine: [String] = []
        for (index, still) in stills.enumerated() {
            let clipStart = index * perImage
            if index > 0, let effectID = transitionEffectID {
                // Centre the dissolve over the cut; stills have unlimited handles.
                spine.append(
                    """
                    <transition name="Cross Dissolve" offset="\(Self.time(frames: clipStart - crossfade / 2, fps: fps))" duration="\(Self.time(frames: crossfade, fps: fps))">
                        <filter-video ref="\(effectID)" name="Cross Dissolve"/>
                    </transition>
                    """
                )
            }
            let name = Self.escape(still.url.deletingPathExtension().lastPathComponent)
            spine.append(
                #"<video ref="\#(assetIDs[index])" offset="\#(Self.time(frames: clipStart, fps: fps))" name="\#(name)" start="0s" duration="\#(Self.time(frames: perImage, fps: fps))"/>"#
            )
        }

        let totalFrames = timing.totalFrames(imageCount: stills.count)
        return """
        <?xml version="1.0" encoding="UTF-8"?>
        <!DOCTYPE fcpxml>
        <fcpxml version="\(Self.version)">
            <resources>
        \(Self.indent(resources, level: 2))
            </resources>
            <library>
                <event name="\(Self.escape(eventName))">
                    <project name="\(Self.escape(projectName))">
                        <sequence format="\(sequenceFormatID)" duration="\(Self.time(frames: totalFrames, fps: fps))" tcStart="0s" tcFormat="NDF">
                            <spine>
        \(Self.indent(spine, level: 7))
                            </spine>
                        </sequence>
                    </project>
                </event>
            </library>
        </fcpxml>

        """
    }

    /// Formats a frame count as an FCPXML rational time, e.g. 60 frames at 30fps → "60/30s".
    public static func time(frames: Int, fps: Int) -> String {
        frames == 0 ? "0s" : "\(frames)/\(fps)s"
    }

    static func sequenceFormatName(width: Int, height: Int, fps: Int) -> String {
        switch (width, height) {
        case (1920, 1080): return "FFVideoFormat1080p\(fps)"
        case (1280, 720): return "FFVideoFormat720p\(fps)"
        default: return "FFVideoFormat\(width)x\(height)p\(fps)"
        }
    }

    static func escape(_ value: String) -> String {
        var result = ""
        result.reserveCapacity(value.count)
        for character in value {
            switch character {
            case "&": result += "&amp;"
            case "<": result += "&lt;"
            case ">": result += "&gt;"
            case "\"": result += "&quot;"
            case "'": result += "&apos;"
            default: result.append(character)
            }
        }
        return result
    }

    private static func indent(_ blocks: [String], level: Int) -> String {
        let pad = String(repeating: "    ", count: level)
        return blocks
            .flatMap { $0.split(separator: "\n", omittingEmptySubsequences: false) }
            .map { pad + $0 }
            .joined(separator: "\n")
    }
}

File: ImageConcat/FcpxmlBuilder.cs
using System.Xml.Linq;

namespace ImgConcat
{
    /// <summary>A still image placed on the Final Cut Pro timeline.</summary>
    public record FcpxmlStill(string Path, int Width, int Height);

    /// <summary>Timeline settings for an FCPXML export. All durations are in whole frames.</summary>
    public record FcpxmlOptions(
        string ProjectName,
        int Width,
        int Height,
        int FrameRate,
        int FramesPerSlide,
        int CrossfadeFrames,
        string EventName = "StopMotion");

    /// <summary>
    /// Builds an FCPXML 1.10 document (Final Cut Pro 10.6 and later) that places each still
    /// on the primary storyline for <see cref="FcpxmlOptions.FramesPerSlide"/> frames, with a
    /// Cross Dissolve centred on every cut when <see cref="FcpxmlOptions.CrossfadeFrames"/> is non-zero.
    /// </summary>
    public static class FcpxmlBuilder
    {
        public const string Version = "1.10";

        // Built-in Final Cut Pro "Cross Dissolve" transition.
        public const string CrossDissolveUid = "FxPlug:4731E73A-8DAC-4113-9A30-AE85B1761265";

        public static string Build(IReadOnlyList<FcpxmlStill> stills, FcpxmlOptions options)
        {
            if (stills.Count == 0)
            {
                throw new ArgumentException("At least one still is required.", nameof(stills));
            }
            if (options.FrameRate <= 0)
            {
                throw new ArgumentOutOfRangeException(nameof(options), "Frame rate must be positive.");
            }
            if (options.FramesPerSlide <= 0)
            {
                throw new ArgumentOutOfRangeException(nameof(options), "Frames per slide must be positive.");
            }

            var fps = options.FrameRate;
            var crossfadeFrames = Math.Clamp(options.CrossfadeFrames, 0, options.FramesPerSlide);
            var resources = new XElement("resources");

            const string sequenceFormatId = "r1";
            resources.Add(new XElement("format",
                new XAttribute("id", sequenceFormatId),
                new XAttribute("name", SequenceFormatName(options.Width, options.Height, fps)),
                new XAttribute("frameDuration", Time(1, fps)),
                new XAttribute("width", options.Width),
                new XAttribute("height", options.Height),
                new XAttribute("colorSpace", "1-1-1 (Rec. 709)")));

            var nextId = 2;
            string? transitionEffectId = null;
            if (crossfadeFrames > 0 && stills.Count > 1)
            {
                transitionEffectId = $"r{nextId++}";
                resources.Add(new XElement("effect",
                    new XAttribute("id", transitionEffectId),
                    new XAttribute("name", "Cross Dissolve"),
                    new XAttribute("uid", CrossDissolveUid)));
            }

            // Stills share a rate-undefined format per distinct size.
            var stillFormats = new Dictionary<(int Width, int Height), string>();
            var assetIds = new List<string>(stills.Count);
            foreach (var still in stills)
            {
                var size = (still.Width, still.Height);
                if (!stillFormats.TryGetValue(size, out var formatId))
                {
                    formatId = $"r{nextId++}";
                    stillFormats[size] = formatId;
                    resources.Add(new XElement("format",
                        new XAttribute("id", formatId),
                        new XAttribute("name", "FFVideoFormatRateUndefined"),
                        new XAttribute("width", still.Width),
                        new XAttribute("height", still.Height),
                        new XAttribute("colorSpace", "1-13-1")));
                }

                var assetId = $"r{nextId++}";
                assetIds.Add(assetId);
                resources.Add(new XElement("asset",
                    new XAttribute("id", assetId),
                    new XAttribute("name", Path.GetFileNameWithoutExtension(still.Path)),
                    new XAttribute("start", "0s"),
                    new XAttribute("duration", "0s"),
                    new XAttribute("hasVideo", "1"),
                    new XAttribute("format", formatId),
                    new XAttribute("videoSources", "1"),
                    new XElement("media-rep",
                        new XAttribute("kind", "original-media"),
                        new XAttribute("src", new Uri(Path.GetFullPath(still.Path)).AbsoluteUri))));
            }

            var spine = new XElement("spine");
            for (int i = 0; i < stills.Count; i++)
            {
                var clipStart = i * options.FramesPerSlide;
                if (i > 0 && transitionEffectId != null)
                {
                    // Centre the dissolve over the cut; stills have unlimited handles.
                    spine.Add(new XElement("transition",
                        new XAttribute("name", "Cross Dissolve"),
                        new XAttribute("offset", Time(clipStart - crossfadeFrames / 2, fps)),
                        new XAttribute("duration", Time(crossfadeFrames, fps)),
                        new XElement("filter-video",
                            new XAttribute("ref", transitionEffectId),
                            new XAttribute("name", "Cross Dissolve"))));
                }

                spine.Add(new XElement("video",
                    new XAttribute("ref", assetIds[i]),
                    new XAttribute("offset", Time(clipStart, fps)),
                    new XAttribute("name", Path.GetFileNameWithoutExtension(stills[i].Path)),
                    new XAttribute("start", "0s"),
                    new XAttribute("duration", Time(options.FramesPerSlide, fps))));
            }

            var totalFrames = stills.Count * options.FramesPerSlide;
            var document = new XDocument(
                new XDeclaration("1.0", "UTF-8", null),
                new XDocumentType("fcpxml", null, null, null),
                new XElement("fcpxml",
                    new XAttribute("version", Version),
                    resources,
                    new XElement("library",
                        new XElement("event",
                            new XAttribute("name", options.EventName),
                            new XElement("project",
                                new XAttribute("name", options.ProjectName),
                                new XElement("sequence",
                                    new XAttribute("format", sequenceFormatId),
                                    new XAttribute("duration", Time(totalFrames, fps)),
                                    new XAttribute("tcStart", "0s"),
                                    new XAttribute("tcFormat", "NDF"),
                                    spine))))));

            using var writer = new Utf8StringWriter();
            document.Save(writer);
            // XDocumentType writes "<!DOCTYPE fcpxml >"; match the form Final Cut Pro exports.
            return writer.ToString().Replace("<!DOCTYPE fcpxml >", "<!DOCTYPE fcpxml>");
        }

        /// <summary>Formats a frame count as an FCPXML rational time, e.g. 60 frames at 30fps → "60/30s".</summary>
        public static string Time(int frames, int frameRate) => frames == 0 ? "0s" : $"{frames}/{frameRate}s";

        private static string SequenceFormatName(int width, int height, int fps) => (width, height) switch
        {
            (1920, 1080) => $"FFVideoFormat1080p{fps}",
            (1280, 720) => $"FFVideoFormat720p{fps}",
            _ => $"FFVideoFormat{width}x{height}p{fps}"
        };

        private sealed class Utf8StringWriter : StringWriter
        {
            public override System.Text.Encoding Encoding => System.Text.Encoding.UTF8;
        }
    }
}
